### PR TITLE
Ebpf 4.0 v25

### PR DESCRIPTION
**Pull request by regit**
Update of #3177 with some patches from @netoptimizer. This is a candidate for merge.

What is provided in this patchset:
- eBPF load balancing: custom load balancing strategy to distribute the data to the sockets
advanced filtering (ippair load balancing is provided)
- eBPF filter: custom filtering at the same level of BPF filter (VLAN filter provided)
- eBPF bypass: bypass for AF_PACKET capture. we discard packets from flow we don't want to see anymore
- XDP bypass: bypass for AF_PACKET capture with drop being done before the datagram reach Linux network stack
- XDP bypass in IPS mode: route packet from one iface to another without passing by the kernel network stack
- XDP cpu redirect: avoid a bottleneck in the socket buffer creation by sharing the load of the creation among selected CPUs (instead of one)

Checkboxes:
- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
- https://redmine.openinfosecfoundation.org/issues/2399
- https://redmine.openinfosecfoundation.org/issues/2289

Changes from #3177 :
- better hash function for the CPU redirect map function
- conditional build of CPU redirect map in the XDP filter
- doc improvement

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output:
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/374
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/157

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -1397,6 +1433,38 @@
...
+    AC_ARG_ENABLE(ebpf,
+	        AS_HELP_STRING([--enable-ebpf],[Enable eBPF support]),
+	        [ enable_ebpf="yes"],
+	        [ enable_ebpf="no"])
+
+    if test "$enable_ebpf" = "yes"; then
+        AC_CHECK_LIB(elf,elf_begin,,LIBELF="no")
+        if test "$LIBELF" = "no"; then
+            echo
+            echo "dead, no elf"
```
Comment (victorjulien) [inline configure.ac]: error msg

File: configure.ac [hunk 2]
```diff
@@ -1397,6 +1433,38 @@
...
+            echo
+            exit 1
+        fi;
+
+        AC_CHECK_LIB(bpf,bpf_object__open,,LIBBPF="no")
+        if test "$LIBBPF" = "no"; then
+            echo
+            echo "dead, no bpf"
```
Comment (victorjulien) [inline configure.ac]: error

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+
+::
+
+ systemctl stop irqbalance
+ systemctl disable irqbalance
+
+Kernel
+~~~~~~
+
+Install kernel 4.13.+ and reboot.
```
Comment (victorjulien) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: seems to be enough to mention running 4.13+ as a requirement?

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+ make && sudo make install
+
+ sudo make install_headers
+ sudo ldconfig
+
+
+Compile and install Suricata
+----------------------------
+
+If ever you don't have the source ::
```
Comment (victorjulien) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: what does this mean?

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+        # completed. If bypass is enabled this will also trigger flow
+        # bypass. If disabled (the default), TLS/SSL session is still
+        # tracked for Heartbleed and other anomalies.
+        no-reassemble: yes
+
+
+Setup eBPF filter
+-----------------
+
+Copy the resulting ebpf fiter as needed ::
```
Comment (victorjulien) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: typo 'fiter

Comment (victorjulien) [reply doc/userguide/capture-hardware/ebpf-xdp.rst]: should the user edit this file too?

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    ebpf-filter-file:  /etc/suricata/ebpf/bypass_filter.bpf
+    bypass: yes
+    use-mmap: yes
+    ring-size: 200000
+
+
+Setup eBPF load balancing
+-------------------------
+
+Copy the resulting ebpf fiter as needed ::
```
Comment (victorjulien) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: typo

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+    ebpf-lb-file:  /etc/suricata/ebpf/lb.bpf
+    use-mmap: yes
+    ring-size: 200000
+
+Setup XDP bypass
+----------------
+
+XDP bypass will allow Suricata to tell the kernel that packets for some
+flows have to be dropped via the XDP mechanism. This is a really early
+drop that occurs before the datagram is reaching the ne
```
Comment (victorjulien) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: 'the ne'

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+also use the ``/etc/suricata/ebpf/xdp_filter.bpf`` (in our example TCP offloading/bypass) ::
+
+  - interface: eth3
+    threads: 16
+    cluster-id: 97
+    cluster-type: cluster_qm # symmetric hashing is a must!
+    defrag: yes
+    # eBPF file containing a 'loadbalancer' function that will be inserted into the
+    # kernel and used as load balancing function
+    #ebpf-lb-file:  /etc/suricata/ebpf/lb.bpf
```
Comment (victorjulien) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: might make sense to leave out the ebpf settings from this example

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+Follow these instructions closely for desired result::
+
+ ifconfig eth3 down
+
+Use in tree kernel drivers: XDP support is not available in Intel drivers available on Intel website.
+
+Enable symmetric hashing::
+
+ ifconfig eth3 down 
+ ethtool -L eth3 combined 16
```
Comment (victorjulien) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: this assumes 16 queues, might be good to mention that this can be different depending on cores?

**Comment by victorjulien** [comment]
it seems that src/ebpf/ doesn't generate anything that becomes part of the suricata binary, so I think I'd prefer to move it down one level

File: src/util-device.c [hunk 10]
```diff
@@ -60,14 +71,48 @@ int LiveGetOffload(void)
...
+
+/**
+ *  \brief Add a pcap device for monitoring and create structure
+ *
+ *  \param dev string with the device name
+ *
+ *  \retval 0 on success.
+ *  \retval -1 on failure.
+ */
+int LiveRegisterDeviceFull(const char *dev)
```
Comment (victorjulien) [inline src/util-device.c]: I don't like this name. It's not self explaining and it's unclear why all callers need it.

Maybe it's better to have a special call for devices that need storage 'RegisterDeviceWithStorage' that are properly set up in Finalize

File: src/util-device.h [hunk 11]
```diff
@@ -50,7 +50,13 @@ typedef struct LiveDevice_ {
     uint32_t offload_orig;  /**< original offload settings to restore @exit */
 } LiveDevice;
 
+typedef struct PreLiveDevice_ {
```
Comment (victorjulien) [inline src/util-device.h]: can we not have 'Pre' prefix? If you have to add something, make it a suffix please

File: suricata.yaml.in [hunk 12]
```diff
@@ -613,12 +613,34 @@ af-packet:
...
+    #  * cluster_ebpf: kernel rotates between sockets following an extended BPF function loaded
+    #  from a file. User needs to use ebpf-lb-file option to set the path to the file
+    #  to load. Requires at least Linux 4.2.
     # Recommended modes are cluster_flow on most boxes and cluster_cpu or cluster_qm on system
     # with capture card using RSS (require cpu affinity tuning and system irq tuning)
     cluster-type: cluster_flow
     # In some fragmentation case, the hash can not be computed. If "defrag" is set
     # to yes, the kernel will do the needed defragmentation before sending the packets.
     defrag: yes
+    # eBPF file containing a 'loadbalancer' function that will be inserted into the
```
Comment (victorjulien) [inline suricata.yaml.in]: I'm inclined to reject this conf update. I think its a too advanced topic to be in the default config.

File: src/util-ebpf.c [hunk 13]
```diff
@@ -0,0 +1,733 @@
...
+ * Flow timeout checking function
+ *
+ * This function is called by the Flow bypass manager to trigger removal
+ * of entries in the kernel/userspace flow table if needed.
+ *
+ */
+int EBPFCheckBypassedFlowTimeout(struct flows_stats *bypassstats,
+                                        struct timespec *curtime)
+{
+    struct flows_stats l_bypassstats = { 0, 0, 0};
```
Comment (victorjulien) [inline src/util-ebpf.c]: l as a leading variable character is evil

File: src/source-af-packet.c [hunk 14]
```diff
@@ -1894,6 +1981,51 @@ int AFPIsFanoutSupported(void)
 #endif
 }
 
+#ifdef HAVE_PACKET_EBPF
+
+static int SockFanoutSeteBPF(AFPThreadVars *ptv)
+{
+    int pfd;
+
+    pfd = ptv->ebpf_lb_fd;
```
Comment (victorjulien) [inline src/source-af-packet.c]: declare var here

File: src/source-af-packet.c [hunk 15]
```diff
@@ -1894,6 +1981,51 @@ int AFPIsFanoutSupported(void)
...
+    SCLogInfo("Activated eBPF on socket");
+
+    return 0;
+}
+
+static int SetEbpfFilter(AFPThreadVars *ptv)
+{
+    int pfd;
+
+    pfd = ptv->ebpf_filter_fd;
```
Comment (victorjulien) [inline src/source-af-packet.c]: declare var

File: src/source-af-packet.c [hunk 16]
```diff
@@ -2112,6 +2263,242 @@ TmEcode AFPSetBPFFilter(AFPThreadVars *ptv)
...
+ * Insert a half flow in the kernel bypass table
+ *
+ * \param mapfd file descriptor of the protocol bypass table
+ * \param key data to use as key in the table
+ * \param inittime time of creation of the entry (in monotonic clock)
+ * \return 0 in case of error, 1 if success
+ */
+static int AFPInsertHalfFlow(int mapd, void *key, uint64_t inittime)
+{
+    unsigned int nr_cpus = UtilCpuGetNumProcessorsConfigured();
```
Comment (victorjulien) [inline src/source-af-packet.c]: would it make sense to run this just once per suri run?

**Comment by victorjulien** [comment]
Pretty minor comments for such a large diff.

I do think we need to be clear on the status of this. While AF_PACKET is a supported feature (tier1), I will consider this eBPF/XDP experimental/community supported until it is tested more and is ironed out more. It does mean that I'm expecting from you that you're maintaining _and_ supporting it.

**Comment by regit** [comment]
@inliniac i'm really good with maintaining it and adding AF_XDP as soon as it is available ;)

File: src/util-ebpf.c (hunk 13 above)
Comment (regit) [reply src/util-ebpf.c]: With a l like leblond :P

File: suricata.yaml.in (hunk 12 above)
Comment (regit) [reply suricata.yaml.in]: You want to remove the full ebpf section and let use use the doc ?

Comment (victorjulien) [reply suricata.yaml.in]: yeah, something like 'for advanced config options for using ebpf and/or xdp see <doc url>'

Comment (victorjulien) [reply suricata.yaml.in]: see _docurl_

File: doc/userguide/capture-hardware/ebpf-xdp.rst (hunk 4 above)
Comment (regit) [reply doc/userguide/capture-hardware/ebpf-xdp.rst]: reworded

**Comment by pevma** [comment]
For the record - I am also good with QAing this (and AF_XDP :) ) on a permanent and regular basis.

File: src/util-device.h (hunk 11 above)
Comment (regit) [reply src/util-device.h]: Renamed to LiveDeviceName as it is that currently

File: src/util-device.c (hunk 10 above)
Comment (regit) [reply src/util-device.c]: Proposal to have the following changes:
- Regular LiveRegisterDevice become LiveRegisterDeviceName
- napatech and pfring use LiveRegisterDevice that do not change meaning

Comment (regit) [reply src/util-device.c]: This gives the following result: https://github.com/regit/suricata/commit/98b621e9424e6f8a336a042cddc48ebc17627de1

Comment (victorjulien) [reply src/util-device.c]: I'm wondering if there is a need for 2 separate tracks here. Can't we just use the same logic for all users? So register only the name from the runmodes and then call the finalize to finish the setup? Right now it wouldn't be very clear to API consumers when to use one or the other.

**Pull request closed without merging**